xarray/core/groupby.py

```diff
@@ -478,35 +480,75 @@ def _infer_concat_args(self, applied_example):
         return coord, dim, positions
 
     def _binary_op(self, other, f, reflexive=False):
+        from .dataarray import DataArray
+        from .dataset import Dataset
```

Comment:
```suggestion
    def _binary_op(self, other: DataArray | Dataset, f, reflexive=False):
```
How about moving the isinstance check to the typing instead?

Reply:
Don't we still want the runtime check?

Reply:
No strong opinion about this really but you can argue that since we typed `other` correctly the runtime check is redundant. Typecheckers will catch wrongly typed inputs, it cleans up the code and it saves some nanoseconds during runtime...

Reply:
I think we keep it, not all of our userbase is using a typechecker.

Feel free to push typing changes to this PR if you'd like.